# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

os.getcwd()

'E:\\Saransh'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv('step3_output.csv')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  LongestRun_mi  SkiableTerrain_ac  \
0          0          2     2       0  ...            1.0             1610.0   
1          0          0     0       0  ...            2.0              640.0   
2          0          0     0       1  ...            1.0               30.0   
3          1          0     2       2  ...            2.0              777.0   
4          0          1     2       3  ...            1.2              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  
0          85.0              150.0           550.0  
1          53.0               90.0             0.0  
2          34.0              152.0            30.0  
3          89.0              122.0             0.0  
4          78.0              104.0            80.0  

[5 rows x 25 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
pd.options.display.max_columns = 100

In [4]:
dfo = df['state']

df1 = pd.concat([df.drop('state',axis = 1),pd.get_dummies(dfo)], axis = 1)
df1.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  surface  total_chairs  Runs  \
0          0          2     2       0       0        2             7  76.0   
1          0          0     0       0       4        0             4  36.0   
2          0          0     0       1       0        2             3  13.0   
3          1          0     2       2       1        2             8  55.0   
4          0          1     2       3       1        0             7  65.0   

   TerrainParks  LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  \
0           2.0            1.0             1610.0           113.0   
1           1.0            2.0              640.0            60.0   
2           1.0            1.0               30.0            30.0   
3           4.0            2.0              777.0           104.0   
4           2.0            1.2              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  Alaska  Arizona  California  Colorado  \
0              150.0           550.0       1        0           0         0   
1               90.0             0.0       1        0           0         0   
2              152.0            30.0       1        0           0         0   
3              122.0             0.0       0        1           0         0   
4              104.0            80.0       0        1           0         0   

   Connecticut  Idaho  Illinois  Indiana  Iowa  Maine  Maryland  \
0            0      0         0        0     0      0         0   
1            0      0         0        0     0      0         0   
2            0      0         0        0     0      0         0   
3            0      0         0        0     0      0         0   
4            0      0         0        0     0      0         0   

   Massachusetts  Michigan  Minnesota  Missouri  Montana  Nevada  \
0              0         0          0         0        0       0   
1              0         0          0         0        0       0   
2              0         0          0         0        0       0   
3              0         0          0         0        0       0   
4              0         0          0         0        0       0   

   New Hampshire  New Jersey  New Mexico  New York  North Carolina  Ohio  \
0              0           0           0         0               0     0   
1              0           0           0         0               0     0   
2              0           0           0         0               0     0   
3              0           0           0         0               0     0   
4              0           0           0         0               0     0   

   Oregon  Pennsylvania  Rhode Island  South Dakota  Tennessee  Utah  Vermont  \
0       0             0             0             0          0     0        0   
1       0             0             0             0          0     0        0   
2       0             0             0             0          0     0        0   
3       0             0             0             0          0     0        0   
4       0             0             0             0          0     0        0

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y , test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [7]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [8]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [9]:
y_pred

array([ 58.30405836,  89.84312086,  78.65366773,  76.35581617,
        69.41733961,  77.03452711,  46.14390211,  47.19468336,
        95.10581617,  84.50523023,  51.62925367,  79.12827711,
        50.90171461,  71.28257398,  63.55405836,  93.59702711,
        40.99351148,  73.08335523,  61.27671461,  40.99546461,
       113.94761304,  35.21812086,  53.67905836,  49.30796461,
        46.60483961,  61.72983961,  77.11655836,  71.69077711,
        64.66929273,  64.75913648,  35.59995679,  59.20249586,
        80.26108961,  57.01792554,  76.08530836,  49.46030836,
        21.45151929,  30.02183179,  63.39390211,  36.41831617,
        46.82749586,  63.46030836,  87.25523023,  72.18491773,
        72.48862867,  59.27671461,  59.30015211, 164.86265211,
        39.31284742,  60.15171461,  70.15855054,  60.61265211,
        53.57554273,  43.54819898,  59.48667554,  42.32358961,
        86.31187086,  38.95444898, 119.34312086, 120.33140211,
        55.07358961,  54.07944898,  42.12437086,  80.90

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model


from sklearn.metrics import explained_variance_score, mean_absolute_error

In [11]:
print('Mean Absolute Error: ', mean_absolute_error(y_test,y_pred))
print('Explained Variance: ', explained_variance_score(y_test,y_pred))

Mean Absolute Error:  4.932204886567597
Explained Variance:  0.9405583157421514


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
print(lm.intercept_)

64.11265210698181


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [13]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        1.522020e+00
vertical_drop      1.076425e+00
trams              2.938641e+12
fastEight          2.888972e+11
fastSixes          3.420095e+12
fastQuads          1.153682e+13
quad               6.886761e+12
triple             8.497317e+12
double             9.525407e+12
surface            1.080913e+13
total_chairs       3.043194e+13
Runs               1.332825e+00
TerrainParks       2.258911e-01
LongestRun_mi      4.965820e-01
SkiableTerrain_ac  2.507324e-01
Snow Making_ac     5.328369e-01
daysOpenLastYear   8.994141e-01
yearsOpen          4.639893e-01
averageSnowfall    1.567383e-01
AdultWeekday       1.934424e+01
projectedDaysOpen  5.078125e-02
NightSkiing_ac     7.504883e-01
Alaska             1.091105e+12
Arizona            8.922445e+11
California         2.806214e+12
Colorado           2.867601e+12
Connecticut        1.404296e+12
Idaho              2.151970e+12
Illinois           1.257971e+12
Indiana            8.922445e+11
Iowa               1.091105e+12
Maine              1.872431e+12
Maryland           6.318732e+11
Massachusetts      2.063591e+12
Michigan           3.254724e+12
Minnesota          2.317072e+12
Missouri           8.922445e+11
Montana            2.151970e+12
Nevada             1.257971e+12
New Hampshire      2.469203e+12
New Jersey         8.922445e+11
New Mexico         1.872431e+12
New York           3.448793e+12
North Carolina     1.535961e+12
Ohio               1.404296e+12
Oregon             1.970638e+12
Pennsylvania       2.677867e+12
Rhode Island       6.318732e+11
South Dakota       8.922445e+11
Tennessee          6.318732e+11
Utah               2.236317e+12
Vermont            2.394599e+12
Virginia           1.257971e+12
Washington         1.970638e+12
West Virginia      1.257971e+12
Wisconsin          2.469203e+12
Wyoming            1.768092e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [14]:
df2 = df.copy('deep')

df2.drop('state', axis = 1, inplace = True)

# Preprocessing data for Model2
X= df2.drop(['Name','AdultWeekend'], axis=1)
y = df2['AdultWeekend']

scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [15]:
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y , test_size=0.25, random_state=1)


In [16]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [17]:
print('Mean Absolute Error: ', mean_absolute_error(y_test,y_pred))
print('Explained Variance: ', explained_variance_score(y_test,y_pred))

Mean Absolute Error:  5.447768043275525
Explained Variance:  0.9267908062622614


In [18]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  

Coefficient
summit_elev           0.932162
vertical_drop         1.726149
trams                 0.263537
fastEight             0.059464
fastSixes             0.440931
fastQuads             0.805663
quad                  1.441450
triple                1.311227
double                0.247775
surface               1.120113
total_chairs          0.702128
Runs                  1.217341
TerrainParks          0.280781
LongestRun_mi         0.314114
SkiableTerrain_ac     0.344249
Snow Making_ac        0.157472
daysOpenLastYear      0.626103
yearsOpen             0.312609
averageSnowfall       1.998401
AdultWeekday         20.213920
projectedDaysOpen     0.236059
NightSkiing_ac        0.456503

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [19]:
df3 = df.copy('deep')

df3.drop(['state','summit_elev'], axis = 1, inplace = True)

# Preprocessing data for Model2
X= df3.drop(['Name','AdultWeekend'], axis=1)
y = df3['AdultWeekend']

scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [20]:
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y , test_size=0.25, random_state=1)


In [21]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [22]:
print('Mean Absolute Error: ', mean_absolute_error(y_test,y_pred))
print('Explained Variance: ', explained_variance_score(y_test,y_pred))

Mean Absolute Error:  5.5688119216459055
Explained Variance:  0.9234199618695252


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.94 | 4.93 |-|
| Model 2. | 0.93| 5.44 |'state'|
| Model 3. | 0.92 | 5.56 |'state','summit_elev','base_elev'|

Model Selection:

**<font color='teal'> Model 1 is chosen as the best model as it performed better than Model 2, 3 while measuring on Explained Varinace and Mean Absolute Error metrics. </font>**